### PR TITLE
chore: bump `@metamask/providers` to `^17.2.0`

### PR DESCRIPTION
**Pull request by jiexi**
## **Description**

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/27049?quickstart=1)

## **Related issues**

- #28060
- #27923

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by jiexi** [comment]
@metamaskbot update-policies

File: development/webpack/utils/loaders/swcLoader.ts [hunk 1]
```diff
@@ -123,12 +127,26 @@ const schema = {
...
+                  type: 'boolean',
+                }
               },
               additionalProperties: false,
               required: ['syntax'],
             },
           ],
         },
+        experimental: {
+          description: 'TODO',
```
Comment (jiexi) [inline development/webpack/utils/loaders/swcLoader.ts:141]: fix this

File: development/webpack/utils/loaders/swcLoader.ts
Comment (jiexi) [inline development/webpack/utils/loaders/swcLoader.ts:1]: these changes probably should be moved into a separate PR

File: development/webpack/utils/loaders/swcLoader.ts [hunk 2]
```diff
@@ -123,12 +127,26 @@ const schema = {
                   default: false,
                   type: 'boolean',
                 },
+                importAssertions: {
+                  default: false,
+                  type: 'boolean',
+                }
```
Comment (jiexi) [inline development/webpack/utils/loaders/swcLoader.ts:133]: DRY this `importAssertions` property somehow

**Comment by jiexi** [comment]
@metamaskbot update-policies

File: development/webpack/utils/loaders/swcLoader.ts [hunk 3]
```diff
@@ -201,6 +223,11 @@ export function getSwcLoader(
         parser: {
           syntax,
           [syntax === 'typescript' ? 'tsx' : 'jsx']: enableJsx,
+          importAssertions: true,
+        },
+        experimental: {
+          keepImportAttributes: true,
+          emitAssertForImportAttributes: true,
```
Comment (jiexi) [inline development/webpack/utils/loaders/swcLoader.ts:230]: unsure if all three or only some are needed, but definitely something needs to be set to deal with the import assertion

**Comment by jiexi** [comment]
```
// node_modules/@metamask/providers/dist/extension-provider/createExternalExtensionProvider.mjs
import { detect } from "detect-browser";
import $extensionportstream from "extension-port-stream";
const { PortDuplexStream: PortStream } = $extensionportstream;
import config from "./external-extension-config.json" assert { type: "json" };
import { MetaMaskInpageProviderStreamName } from "../MetaMaskInpageProvider.mjs";
import { StreamProvider } from "../StreamProvider.mjs";
import { getDefaultExternalMiddleware } from "../utils.mjs";
const browser = detect();
/**
 * Creates an external extension provider for the given extension type or ID.
 *
 * @param typeOrId - The extension type or ID.
 * @returns The external extension provider.
 */
export function createExternalExtensionProvider(typeOrId = 'stable') {
    let provider;
    try {
        const extensionId = getExtensionId(typeOrId);
        const metamaskPort = chrome.runtime.connect(extensionId);
        const pluginStream = new PortStream(metamaskPort);
        provider = new StreamProvider(pluginStream, {
            jsonRpcStreamName: MetaMaskInpageProviderStreamName,
            logger: console,
            rpcMiddleware: getDefaultExternalMiddleware(console),
        });
        // This is asynchronous but merely logs an error and does not throw upon
        // failure. Previously this just happened as a side-effect in the
        // constructor.
        // eslint-disable-next-line @typescript-eslint/no-floating-promises
        provider.initialize();
    }
    catch (error) {
        console.dir(`MetaMask connect error.`, error);
        throw error;
    }
    return provider;
}
/**
 * Gets the extension ID for the given extension type or ID.
 *
 * @param typeOrId - The extension type or ID.
 * @returns The extension ID.
 */
function getExtensionId(typeOrId) {
    const ids = browser?.name === 'firefox' ? config.firefoxIds : config.chromeIds;
    return ids[typeOrId] ?? typeOrId;
}
//# sourceMappingURL=createExternalExtensionProvider.mjs.map
```

File: development/webpack/utils/loaders/swcLoader.ts (hunk 3 above)
Comment (jiexi) [reply development/webpack/utils/loaders/swcLoader.ts:230]: `import config from "./external-extension-config.json" assert { type: "json" };`

File: .yarn/patches/@metamask-providers-npm-17.2.0-cb0cfcdadc.patch [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+index f2aa9285d84cd5e25c20f5202437c476cdacc326..121a9099e456d3444d8aae79574ba9c53e7f2699 100644
+--- a/dist/extension-provider/createExternalExtensionProvider.mjs
++++ b/dist/extension-provider/createExternalExtensionProvider.mjs
+@@ -1,5 +1,4 @@
+-import $detectbrowser from "detect-browser";
+-const { detect } = $detectbrowser;
++import { detect } from "detect-browser";
+ import $extensionportstream from "extension-port-stream";
+ const { PortDuplexStream: PortStream } = $extensionportstream;
+ import config from "./external-extension-config.json" assert { type: "json" };
```
Comment (jiexi) [inline .yarn/patches/@metamask-providers-npm-17.2.0-cb0cfcdadc.patch:11]: without this patch, I get the following from the inpage provider in the dapp web console

```
createExternalExtensionProvider.mjs:2 Uncaught TypeError: Cannot destructure property 'detect' of 'detect_browser__WEBPACK_IMPORTED_MODULE_0__.default' as it is undefined.
    at 193000 (createExternalExtensionProvider.mjs:2:9)
    at __webpack_require__ (bootstrap:19:1)
    at 701043 (createExternalExtensionProvider.mjs:49:61)
    at __webpack_require__ (bootstrap:19:1)
    at make namespace object:7:1
    at inpage.js:70:1
    at inpage.js:70:1
 ```

File: app/scripts/inpage.js [hunk 5]
```diff
@@ -34,7 +34,7 @@ cleanContextForImports();
 import log from 'loglevel';
 import { v4 as uuid } from 'uuid';
 import { WindowPostMessageStream } from '@metamask/post-message-stream';
-import { initializeProvider } from '@metamask/providers/dist/initializeInpageProvider';
```
Comment (MajorLift) [inline app/scripts/inpage.js:37]: Relevant: https://github.com/MetaMask/snaps/pull/2445#discussion_r1678337887

File: app/scripts/inpage.js [hunk 6]
```diff
@@ -34,7 +34,7 @@ cleanContextForImports();
...
+import { initializeProvider } from '@metamask/providers';
```
Comment (legobeat) [inline app/scripts/inpage.js:37]:
```suggestion
import { initializeProvider } from '@metamask/providers/dist/initializeInpageProvider.cjs';
```

File: package.json [hunk 7]
```diff
@@ -267,7 +267,9 @@
     "@metamask/nonce-tracker@npm:^5.0.0": "patch:@metamask/nonce-tracker@npm%3A5.0.0#~/.yarn/patches/@metamask-nonce-tracker-npm-5.0.0-d81478218e.patch",
     "@metamask/keyring-controller@npm:^16.0.0": "patch:@metamask/keyring-controller@npm%3A17.1.1#~/.yarn/patches/@metamask-keyring-controller-npm-17.1.1-098cb41930.patch",
     "@metamask/keyring-controller@npm:^17.1.0": "patch:@metamask/keyring-controller@npm%3A17.1.1#~/.yarn/patches/@metamask-keyring-controller-npm-17.1.1-098cb41930.patch",
-    "@trezor/connect-web@npm:^9.1.11": "patch:@trezor/connect-web@npm%3A9.3.0#~/.yarn/patches/@trezor-connect-web-npm-9.3.0-040ab10d9a.patch"
+    "@trezor/connect-web@npm:^9.1.11": "patch:@trezor/connect-web@npm%3A9.3.0#~/.yarn/patches/@trezor-connect-web-npm-9.3.0-040ab10d9a.patch",
+    "@metamask/providers@npm:^17.1.2": "patch:@metamask/providers@npm%3A17.2.0#~/.yarn/patches/@metamask-providers-npm-17.2.0-cb0cfcdadc.patch",
+    "@metamask/providers@npm:^15.0.0": "patch:@metamask/providers@npm%3A17.2.0#~/.yarn/patches/@metamask-providers-npm-17.2.0-cb0cfcdadc.patch"
```
Comment (legobeat) [inline package.json:272]: Is it safe to force usage expecting v15 to use v17 instead?

**Comment by jiexi** [comment]
@metamaskbot update-policies

**Comment by Gudahtt** [comment]
Replaced by https://github.com/MetaMask/metamask-extension/pull/28757
